File: lab3/cpp/test/test_dh_user.cpp
#include "diffie_hellman.hpp"

#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest.h>

#include <tuple>

using gf1234567891 = jpp::galois_field<1234567891>;
using sut_type = jpp::dh_user<gf1234567891>;
using dh_setup_type = jpp::dh_setup<gf1234567891>;

namespace {

dh_setup_type setup;

gf1234567891 key = 111;
gf1234567891 message = 123;
gf1234567891 code = 321;

} // namespace

TEST_CASE("encrypt should throw if the private key has not been set") {
    sut_type sut(setup);
    REQUIRE_THROWS_AS(std::ignore = sut.encrypt(message), std::logic_error);
}

TEST_CASE("encrypt should return an encrypted message if a private key is set") {
    sut_type sut(setup);
    sut.set_key(key);

    REQUIRE_NOTHROW(std::ignore = sut.encrypt(message));
}

TEST_CASE("decrypt should throw if the private key has not been set") {
    sut_type sut(setup);
    REQUIRE_THROWS_AS(std::ignore = sut.decrypt(code), std::logic_error);
}

TEST_CASE("decrypt should return an encrypted message if a private key is set") {
    sut_type sut(setup);
    sut.set_key(key);

    REQUIRE_NOTHROW(std::ignore = sut.decrypt(code));
}

File: lab3/python/test/test_dh_user.py
import pytest

from dh.common import GF1234567891, GF1234577
from dh.dh_setup import dh_setup_builder
from dh.dh_user import dh_user_builder


DHSetup = dh_setup_builder(GF1234567891)
DHUser = dh_user_builder(GF1234567891)


def test_initialization_exception():
    InvalidDHSetup = dh_setup_builder(GF1234577)

    with pytest.raises(TypeError):
        _ = DHUser(InvalidDHSetup())


setup = DHSetup()
key = GF1234567891(111)
message = GF1234567891(123)
code = GF1234567891(321)


def test_is_compatible():
    sut = DHUser(setup)

    assert sut.is_compatible(GF1234567891)
    assert not sut.is_compatible(GF1234577)


def test_encrypt_exception():
    sut = DHUser(setup)

    with pytest.raises(ValueError):
        _ = sut.encrypt(message)


def test_encrypt():
    sut = DHUser(setup)
    sut.set_key(key)

    assert isinstance(sut.encrypt(message), GF1234567891)


def test_decrypt_exception():
    sut = DHUser(setup)

    with pytest.raises(ValueError):
        _ = sut.decrypt(code)


def test_decrypt():
    sut = DHUser(setup)
    sut.set_key(key)

    assert isinstance(sut.decrypt(code), GF1234567891)
